base/compiler/ssair/EscapeAnalysis/EAUtils.jl

```diff
@@ -0,0 +1,413 @@
+   ↑  11 ─ %27 = Base.getfield(%24, :x)::String                           │││╻   getproperty
+   ◌  └───       goto #13                                                 ││
+   ◌  12 ─       Main.throw(%24)::Union{}                                 ││
+   ◌  └───       unreachable                                              ││
+12 ↑  13 ─ %31 = Base.getfield(%23, :x)::String                           │╻╷╷ get′
+13 ↑  │    %32 = Core.tuple(%27, %31)::Tuple{String, String}              │
+   ◌  └───       return %32                                               │
+```
+
+The symbols in the side of each call argument and SSA statements represents the following meaning:
+- `◌`: this value is not analyzed because escape information of it won't be used anyway (when the object is `isbitstype` for example)
+- `✓`: this value never escapes (`has_no_escape(result.state[x])` holds)
+- `↑`: this value can escape to the caller via return (`has_return_escape(result.state[x])` holds)
+- `X`: this value can escape to somewhere the escape analysis can't reason about like escapes to a global memory (`has_all_escape(result.state[x])` holds)
+- `*`: this value's escape state is between the `ReturnEscape` and `AllEscape` in the `EscapeLattice`, e.g. it has unhandled `ThrownEscape`
+and additional `′` indicates that field analysis has been done successfully on that value.
+
+For testing, escape information of each call argument and SSA value can be inspected programmatically as like:
+```julia
+julia> result.state[Core.Argument(3)]
+ReturnEscape
+
+julia> result.state[Core.SSAValue(3)]
+NoEscape′
+```
+"""
+function code_escapes(@nospecialize(args...);
+                      world = get_world_counter(),
+                      interp = Core.Compiler.NativeInterpreter(world))
+    interp = EscapeAnalyzer(interp)
+    results = code_typed(args...; optimize=true, world, interp)
+    isone(length(results)) || throw(ArgumentError("`code_escapes` only supports single analysis result"))
+    return EscapeResult(interp.ir, interp.state, interp.linfo)
+end
+
+# AbstractInterpreter
+# -------------------
+
+# imports
+import .CC:
+    AbstractInterpreter,
+    NativeInterpreter,
+    WorldView,
+    WorldRange,
+    InferenceParams,
+    OptimizationParams,
+    get_world_counter,
+    get_inference_cache,
+    lock_mi_inference,
+    unlock_mi_inference,
+    add_remark!,
+    may_optimize,
+    may_compress,
+    may_discard_trees,
+    verbose_stmt_info,
+    code_cache,
+    @timeit,
+    get_inference_cache,
+    convert_to_ircode,
+    slot2reg,
+    compact!,
+    ssa_inlining_pass!,
+    sroa_pass!,
+    adce_pass!,
+    type_lift_pass!,
+    JLOptions,
+    verify_ir,
+    verify_linetable
+# usings
+import Core:
+    CodeInstance, MethodInstance, CodeInfo
+import .CC:
+    OptimizationState, IRCode
+import .EA:
+    analyze_escapes, cache_escapes!
+
+mutable struct EscapeAnalyzer{State} <: AbstractInterpreter
```

Comment:
Is this AbsInt necessary to have in Base?

Reply:
No, we can define EA-specific entry within `Core.Compiler` and EAUtils.jl can just consume it.
Do you have any concern related to having an external AbsInt in Base?